### PR TITLE
Recompute sector bounds when nodes are transformed

### PR DESCRIPTION
**Pull request by eiriklegernaes**
#### Type of change

Bug 
(or feature?)
### Overview
When model nodes are transformed (using `CogniteCadModel.setNodeTransform`) they might end up outside the original sector bounds. This in turn may lead to a situation where a transformed node is culled erroneously, because the original sector bounds are outside the view frustum. This PR fixes this problem by dynamically growing the sector bounds to account for any transformed nodes in the sector. This assumes the number of transformed nodes are kept relatively low. If every node is transformed a lot, the sector tree will not be very efficient. That being said, the problem of disappearing model nodes is a correctness problem, and it's trivial to encounter. I therefore think this trade off is acceptable.

### Details
Before sector bounds can be changed, it is necessary to know which sector a certain tree index has geometry in. In the future, this could be pre-computed and accessed through the backend, but as of now this is done by iterating over the sector data when a sector is first loaded. This is stored in a map, and only done once per sector per geometry type. 

A `Cognite3DModel` now has an instance of the `CustomSectorBounds` class. When this class is given the following info
- The tree index of the transformed node
- The transform
- The set of sectors node has geometry in

it is able to adjust the sector bounds so that each sector still contains the transformed nodes. Both the transform and the set of sectors may be updated, meaning this book-keeping works even when transforming a node before any sectors are loaded. When transformed nodes are reset back to their original transform, the sector bounds will shrink back to the original value if possible.

### Performance impact

The performance impact of the `TreeIndexToSectorsMap` book-keeping appears to be low, but it's hard to supply conclusive evidence. I've measured the time it takes to loop over the data in a sector, and it's usually roughly around ~100us (sometimes probably due to timer resolution). There are many sectors, but the impact doesn't appear to be large relative to other operations which are performed on each sector. 

I've also attempted to measure the memory footprint of this map, as it currently doesn't employ any bounding mechanism. After navigating through a medium sized asset for about 10 minutes, I've measured it to be about 30 MB. This is about 10% of the total footprint, with the majority being buffers with what I assume is sector data (`InstancedInterleavedBuffer` and `BufferAttribute` arrays). A sophisticated bounding scheme could be used, dropping entries for tree indices in sectors haven't been loaded in a long time, but this would require more book-keeping.

The only public facing change should be that `CogniteCadModel.setNodeTransform` is now `async`.

## How has this been tested? :mag:
Both classes have accompanying tests. The non-trivial logic in `CustomSectorBounds` is tested on a mocked up sector tree, and should hopefully give confidence that the logic is correct. It's also manually tested in the example viewer, by modifying it to translate nodes on click, and visualizing the sector bounds, but it's very hard to test edge cases this way.

### CODE REVIEW

**Review by haakonflatval-cognite — CHANGES_REQUESTED** [review]
Very nice feature! Thanks for the contribution!

I've made a few comments regarding code style, but otherwise I'm very happy with this. 
I've made some testing on the impact of the extra tree-index-sector registration done on loading times, and I've seen about 3-6% lower FPS during the test run. But I'd say that's within tolerance.

File: viewer/packages/cad-model/src/utilities/TreeIndexToSectorsMap.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Map between a tree index and the set of sectors it has geometry in. Also contains helper functions to keep track of whether or not
+ * a given sector has had all its tree indices added to the map. This means iterating over all tree indices in a sector will only be
+ * done on first load.
+ */
+export class TreeIndexToSectorsMap {
+  public onChange?: (treeIndex: number, newSectorId: number) => void;
+  private readonly treeIndexToSectorIds = new Map<number, Set<number>>();
+  private readonly parsedSectors = new Map<number, Set<RevealGeometryCollectionType>>();
```
[the remark this answers is not in the document]
Comment (eiriklegernaes) [reply viewer/packages/cad-model/src/utilities/TreeIndexToSectorsMap.ts]: Fixed!

File: viewer/packages/cad-model/src/utilities/TreeIndexToSectorsMap.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    if (existingSet) {
+      if (!existingSet.has(sectorId)) {
+        existingSet.add(sectorId);
+        this.onChange?.(treeIndex, sectorId);
+      }
+    } else {
+      this.treeIndexToSectorIds.set(treeIndex, new Set<number>([sectorId]));
+      this.onChange?.(treeIndex, sectorId);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (eiriklegernaes) [reply viewer/packages/cad-model/src/utilities/TreeIndexToSectorsMap.ts]: Done!

File: viewer/packages/cad-geometry-loaders/src/CadManager.ts [hunk 3]
```diff
@@ -102,6 +103,19 @@ export class CadManager {
...
+        if (sector.group?.children.length === 1) {
+          const treeIndices = sector.group.children[0].userData?.treeIndices as Map<number, number> | undefined;
+          if (treeIndices !== undefined) {
+            for (const treeIndex of treeIndices.keys()) {
+              cadModel.treeIndexToSectorsMap.set(treeIndex, sector.metadata.id);
+            }
+            cadModel.treeIndexToSectorsMap.markCompleted(sector.metadata.id, RevealGeometryCollectionType.TriangleMesh);
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (eiriklegernaes) [reply viewer/packages/cad-geometry-loaders/src/CadManager.ts]: Fixed!

File: viewer/packages/cad-model/src/utilities/TreeIndexToSectorsMap.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+  /**
+   * Get the set of sectors a tree index is known to have geometry in
+   * @param treeIndex Tree index
+   * @returns The set of sectors
+   */
+  get(treeIndex: number): Set<number> {
```
[the remark this answers is not in the document]
Comment (eiriklegernaes) [reply viewer/packages/cad-model/src/utilities/TreeIndexToSectorsMap.ts]: Fixed

File: viewer/packages/cad-model/src/utilities/CustomSectorBounds.ts [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    const transformedNode = this._treeIndexToTransformedNodeMap.get(treeIndex);
+    if (!transformedNode) {
+      throw new Error(`Attempted to update node sectors for non-registered node (tree index ${treeIndex})`);
+    }
+
+    for (const newSector of newSectors) {
+      if (!transformedNode.inSectors.has(newSector)) {
+        // Add sector to transformed node
+        transformedNode.inSectors.add(newSector);
+
```
[the remark this answers is not in the document]
Comment (eiriklegernaes) [reply viewer/packages/cad-model/src/utilities/CustomSectorBounds.ts]: Fixed!

File: viewer/packages/cad-model/src/utilities/TreeIndexToSectorsMap.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    return this.treeIndexToSectorIds.get(treeIndex) ?? new Set<number>();
+  }
+
+  /**
+   * Mark a sector as completed for a given geometry type. This will make subsequent calls to isCompleted
+   * for this geometry type return true
+   * @param sectorId The sector id
+   * @param type The geometry type
+   */
+  markCompleted(sectorId: number, type: RevealGeometryCollectionType): void {
```
[the remark this answers is not in the document]
Comment (eiriklegernaes) [reply viewer/packages/cad-model/src/utilities/TreeIndexToSectorsMap.ts:51]: Leaving it as it is, as we discussed, to avoid having to build an array of tree indices at the call sites.

**Review by haakonflatval-cognite — APPROVED** [review]
